### ISSUE TITLE
Is this actually BM25?

### ISSUE THREAD

**Issue by mdagost** [pinecone-io/pinecone-text#69]
I don't think the implementation here for the `BM25Encoder` is standard BM25.

If you have a look at the standard B25 formula from, e.g. [Wikipedia](https://en.wikipedia.org/wiki/Okapi_BM25), it has two terms in the sum--the IDF term and the other term.  The other term is what is being calculated during the document encoding phase [here](https://github.com/pinecone-io/pinecone-text/blob/main/pinecone_text/sparse/bm25_encoder.py#L125-L127).  It leaves out the constant k+1 factor but that shouldn't matter for the ranking when the dot product happens with the query later.

The IDF term appears to be calculated via the query encoding [here](https://github.com/pinecone-io/pinecone-text/blob/main/pinecone_text/sparse/bm25_encoder.py#L155-L157).

There are two issues with this calculation.  The first is that the variable `tf` [here](https://github.com/pinecone-io/pinecone-text/blob/main/pinecone_text/sparse/bm25_encoder.py#L155) is confusingly misnamed because that's actually document frequency, not term frequency.

Second, the formula is just different from the usual BM25 formula.  What is implemented here appears to be, using the Wikipedia notation,

```
ln { (N + 1) / (n(q) + 0.5) }
```

It's also then normalized.  The BM25 IDF term is

```
ln { 1 + (N - n(q) + 0.5) / (n(q) + 0.5) }
```

and it isn't normalized.

Are you all sure that your implementation is correct?  Or did you intend to implement some other BM25?  If it's the latter could you include a reference to it?

**Comment by acatav** [comment]
Hi @mdagost,

- Regarding the naming - you are completely right I created a PR to fix it: #70 
- Regarding the IDF term - note that both terms are identical: `ln { 1 + (N - n(q) + 0.5) / (n(q) + 0.5) } = ln { (n(q) + 0.5) / (n(q) + 0.5) + (N - n(q) + 0.5) / (n(q) + 0.5)} = ln { ( N + n(q) - n(q) + 0.5 + 0.5) } = ln { (N + 1) / (n(q) + 0.5) }`

**Comment by mdagost** [comment]
Doh. Sorry for the noise over some dumb algebra :)
